src/data/finset/noncomm_prod.lean

```diff
@@ -0,0 +1,158 @@
+/-
+Copyright (c) 2021 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+
+import data.finset.fold
+
+/-!
+# Products (respectively, sums) over a finset or a multiset.
+
+The regular `finset.prod` and `multiset.prod` require `[comm_monoid α]`.
+Often, there are collections `s : finset α` where `[monoid α]` and we know,
+in a dependent fashion, that for all the terms `∀ (x ∈ s) (y ∈ s), commute x y`.
+This allows to still have a well-defined product over `s`.
+
+## Main definitions
+
+- `finset.noncomm_prod`, requiring a proof of commutativity of held terms
+- `multiset.noncomm_prod`, requiring a proof of commutativity of held terms
+
+## Implementation details
+
+The commutativity proof required has a somewhat unnecessary `x ≠ y` hypothesis.
+It is unnecesary because `commute x x` is true by reflexivity. The `x ≠ y` hypothesis
+is included to facilitate usage with `set.pairwise_on`, or in helping with implications
+that deal with `disjoint`.
+
+-/
+
+variables {α : Type*} [monoid α]
+
+namespace multiset
+
+/-- Product of a `s : multiset α` with `[monoid α]`, given a proof that `*` commutes
+on all elements `x ∈ s`. -/
+@[to_additive "Sum of a `s : multiset α` with `[add_monoid α]`, given a proof that `*` commutes
+on all elements `x ∈ s`." ]
+def noncomm_prod : Π (s : multiset α) (comm : ∀ (x ∈ s) (y ∈ s) (h : x ≠ y), commute x y), α :=
+λ s, quotient.hrec_on s (λ l h, l.prod)
+  (λ l l' (h : l ~ l'),
+    begin
+      ext,
+      { refine forall_congr (λ x, _),
+        refine imp_congr h.mem_iff _,
+        refine forall_congr (λ y, _),
+        exact imp_congr h.mem_iff iff.rfl },
+      { intros hl hl' _,
+        rw heq_iff_eq,
+        refine h.prod_eq' _,
+        rw list.pairwise.imp_mem,
+        refine list.pairwise_of_forall _,
+        intros x y hx hy,
+        by_cases hxy : x = y,
+        { subst hxy },
+        { exact hl x hx y hy hxy } }
+    end)
```

Comment:
Can you generalize this to `multiset.fold_noncomm`?

Reply:
Not easily, because `multiset.fold` is in terms of `multiset.foldr`, which both require commutativity of the underlying function. I am trying to do it via `multiset.attach` currently.

Reply:
Done.